### ISSUE TITLE
Add support for magic tap

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#3816]
iOS accessibility has the notion of a context-sensitive "perform the action that makes the most sense here" action -- they call it the Magic Tap.

Per the [iOS Developer Library](https://developer.apple.com/library/ios/featuredarticles/ViewControllerPGforiPhoneOS/SupportingAccessibility.html):

> Magic Tap. A two-finger double-tap that performs the most-intended action.

This can be used, for example, to allow the user to play and pause music in a music app without hunting for the play/pause button.

**Comment by chinmaygarde** [comment]
Can we combine all the remaining accessibility TODOs into a single ticket with a checklist please?

**Comment by ddfreiling** [comment]
@tvolkert @jonahwilliams I'm trying to implement `accessibilityPerformMagicTap` in a generalized way like `SemanticsAction.onDismiss` implemented `accessibilityPerformEscape`.

Could you look at the questions I have on https://github.com/ddfreiling/flutter/pull/1 to help me finish up this feature?

Thanks in advance!

**Comment by tvolkert** [comment]
@ddfreiling I didn't see your comment or PRs since it was right in the thick of lockdown.  Sorry about that!  I see you closed the engine PR - was it just because of lack of reviews?

**Comment by ddfreiling** [comment]
Hi @tvolkert, no worries thanks for finally reaching out to me.

This became apparent as we in https://nota.dk service the blind and visually impaired. We would like to make the next version of our app in Flutter, but ran into some missing a11y features, that we were used to just working out of the box, like the a11y gestures MagicTap and Escape.

The reason for closing the engine issue, is that I could not solve the problems documented in the "Questions for Flutter Team" in https://github.com/ddfreiling/flutter/pull/1 .

I will paste them here as well as they are still very relevant questions.

## Questions to the Flutter Team
I'm new to the SDK and trying my best, but need some guidance here:
1. Why does the `Semantics` widget become focusable with VoiceOver whenever an action is added to it? Example: Semantics with props `container: false, onDismiss: () {}` is focusable as an outer node.
   * This makes it impossible to set an onDismiss action for the entire Scaffold, as it then becomes a focusable a11y node.
   * This currently forces us to wrap all back buttons in a Semantics widget with an `onDismiss` callback. 
2. Why does it seem like `SemanticsAction.onDismiss` "bubbles" upwards in the semantics tree, but `onDefaultAction` which I implemented does not. E.g. wrapping the Scaffold in a Semantics node which handles `onDismiss` and `onDefaultAction` only works for `onDismiss`.
   * I've ended up implementing `accessibilityPerformMagicTap` in the AppDelegate and notifying the app, but I still think there should be a Semantics action for this.
3. What is the best approach for defining an app-level or route-level magic-tap og dismiss handler? Normally this would be done via the AppDelegate, but in Flutter I guess it should be on the Scaffold or MaterialApp?

**Comment by tvolkert** [comment]
@ddfreiling, thanks for all the info.

@chunhtai can you have a look?

**Comment by ddfreiling** [comment]
> @ddfreiling, thanks for all the info.
> 
> 
> 
> @chunhtai can you have a look?

Would appreciate some feedback here, so we can have full accessibility support in Flutter.

**Comment by chunhtai** [comment]
Hi @ddfreiling, thanks for contribution! 

I can answer those questions.
> 1. Why does the Semantics widget become focusable with VoiceOver whenever an action is added to it

The semantics node must return true for `isAccessibilityElement` in order to perform `accessibilityPerformEscape` on the node. This will also cause it to be focusable. I am not aware of an accessibility api in iOS to get around this problem. One workaround will be making it return false isAccessibilityElement but manually wires up the children to delegate `accessibilityPerformEscape` to this node

> 2. Why does it seem like SemanticsAction.onDismiss "bubbles" upwards in the semantics tree.

We are not in control of whether the `accessibilityPerformEscape` will bubble up, it is the iOS accessibility service that decide to call `accessibilityPerformEscape` up the tree until one of the node return a YES

Similarly, we do not have control of whether `accessibilityPerformMagicTap` will bubble up the tree. We can work around this by manually write the logic to bubble up the tree in accessibility bridge. Have you check what is the native behavior on this? if you set accessibilityPerformMagicTap for a parent node, will performing the magic tap of its children actually triggers the accessibilityPerformMagicTap on the parent?

> 3. What is the best approach for defining an app-level or route-level MagicTap or Escape handler.

I think it should be on the modal route class to the deal with the escape, but that means we might need to do something to make it work without making the whole route focusable

**Comment by ddfreiling** [comment]
@chunhtai I understand the reasoning for the 2 first points and can accept the limitations.

As for where we go from here.. I would think that you on the Flutter team will need to find a solution to the route-level semantics escape action, without making it focusable.

I still think there should be an "in-SDK" way to set an `accessibilityPerformMagicTap` action on an app or route-level. In my previously referenced pull-request it was simply a new action on the Semantics widget, but due to the above "bubble-up" limitations, it probably makes most sense be able to set it on app or route-level. I could implement this, but would still run into the "focusable" issue, so until this is resolved I will await the Flutter team's decision.

**Comment by chunhtai** [comment]
@ddfreiling I agree we should do something with the focusable problem, it feels to me that onDismiss should not treated as regular semantics action in iOS.

For `accessibilityPerformMagicTap`, can it also be set in route level or app level in native iOS application? The fact that it does not bubble up in iOS accessiblity make me think it is specific to one semantics node.

**Comment by ddfreiling** [comment]
> @ddfreiling I agree we should do something with the focusable problem, it feels to me that onDismiss should not treated be as regular semantics action that is tie to one semantics node.
> 
> For `accessibilityPerformMagicTap`, can it also be set in route level or app level in native iOS application? The fact that it does not bubble up in iOS accessiblity make me think it is specific to one semantics node.

In standard iOS apps `accessibilityPerformMagicTap` can be overridden in AppDelegate or any UIViewController. It will perform the `accessibilityPerformMagicTap` defined in the current UIView and/or AppDelegate, depending on if the implemented function returns true (handled). I am unsure if it actually "bubbles up" if not defined in the nearest UIViewController, but it definitely always bubbles to AppDelegate if not handled further down in the UI tree.

See https://developer.apple.com/library/archive/featuredarticles/ViewControllerPGforiPhoneOS/SupportingAccessibility.html

**Comment by vsomayaji** [comment]
We're handling [`accessibilityPerformMagicTap`](https://developer.apple.com/documentation/objectivec/nsobject/1615137-accessibilityperformmagictap) in `AppDelegate` and notifying the client using a [`FlutterEventChannel`](https://api.flutter.dev/objcdoc/Classes/FlutterEventChannel.html). I'm hoping there will eventually be an easier way for developers to do this, as the Magic Tap is an important gesture for VoiceOver users. I liked the direction @ddfreiling was heading, but it looks like that effort stalled. ☹️ 

<details><summary>AppDelegate.swift</summary>

```swift
import Flutter
import UIKit

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  private var eventSink: FlutterEventSink?

  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    let controller: FlutterViewController = window?.rootViewController as! FlutterViewController
    let eventChannel = FlutterEventChannel(
      name: "aira.io/accessibilityPerformMagicTap",
      binaryMessenger: controller.binaryMessenger)
    eventChannel.setStreamHandler(self)

    GeneratedPluginRegistrant.register(with: self)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }

  // If the VoiceOver two-finger double-tap (i.e. the Magic Tap) is performed, send an event to the
  // aira.io/accessibilityPerformMagicTap channel.
  override func accessibilityPerformMagicTap() -> Bool {
    guard let eventSink = eventSink else {
      return false
    }
    eventSink("accessibilityPerformMagicTap")
    return true
  }
}

extension AppDelegate: FlutterStreamHandler {
  func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
}
```
</details>

<details><summary>main.dart</summary>

```dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/services.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();

  // Auto-enable accessibility for our Blind and Low Vision customers.
  RendererBinding.instance.setSemanticsEnabled(true);

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  static const String _title = 'Magic Tap Demo';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(
          title: const Text(_title),
        ),
        body: const MyStatefulWidget(),
      ),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  const MyStatefulWidget({super.key});

  @override
  State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  bool _playing = false;
  StreamSubscription? _magicTapSubscription;

  @override
  void dispose() {
    _magicTapSubscription?.cancel();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();

    if (defaultTargetPlatform == TargetPlatform.iOS) {
      EventChannel magicTapEventChannel = const EventChannel('aira.io/accessibilityPerformMagicTap');
      _magicTapSubscription = magicTapEventChannel.receiveBroadcastStream().listen((_) {
        _togglePlaying();
        SemanticsService.announce(_playing ? 'Playing' : 'Paused', TextDirection.ltr);
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          children: <Widget>[
            ElevatedButton(
              onPressed: _togglePlaying,
              child: Text(_playing ? 'Pause' : 'Play'),
            ),
          ],
        ),
      ),
    );
  }

  void _togglePlaying() {
    setState(() {
      _playing = !_playing;
    });
  }
}
```
</details>

**Comment by ddfreiling** [comment]
@vsomayaji This is still how we do 2-finger dismiss/scrub as well. I may look into my proposed implementation again at some point and see if the Flutter semantics propagation works any better now.
